File: frontend/src/mocks/handlers/status.js
import { http, HttpResponse } from 'msw';

const services = [
  { name: 'PostgreSQL', slug: 'postgres', status: 'operational', uptime30d: '99.99%', latencyMs: 7 },
  { name: 'Redis', slug: 'redis', status: 'operational', uptime30d: '99.98%', latencyMs: 2 },
  { name: 'Web App', slug: 'web', status: 'operational', uptime30d: '99.95%' },
  { name: 'REST API', slug: 'api', status: 'operational', uptime30d: '99.96%' },
  { name: 'WebSocket', slug: 'ws', status: 'degraded', uptime30d: '99.50%' },
  { name: 'MinIO', slug: 'minio', status: 'operational', uptime30d: '100.00%' },
  { name: 'Judge0', slug: 'judge0', status: 'operational', uptime30d: '99.90%' },
  { name: 'OpenRouter', slug: 'openrouter', status: 'operational', uptime30d: '99.97%' },
];

function fakeBuckets(slug, days) {
  const out = [];
  const today = new Date();
  for (let i = days - 1; i >= 0; i--) {
    const d = new Date(today);
    d.setUTCDate(d.getUTCDate() - i);
    const day = d.toISOString().slice(0, 10);
    let status = 'operational';
    if (slug === 'ws' && i < 3) status = 'degraded';
    if (slug === 'judge0' && i === 12) status = 'down';
    out.push({ day, status });
  }
  return out;
}

export const statusHandlers = [
  http.get('/api/v1/status', () =>
    HttpResponse.json({
      overallStatus: 'degraded',
      uptime90d: '99.92%',
      services,
      incidents: [
        {
          id: 'inc-001',
          title: 'WebSocket задержки',
          description: 'Часть соединений долго отдают первый pong.',
          severity: 'minor',
          startedAt: new Date(Date.now() - 2 * 3600_000).toISOString(),
          endedAt: null,
          affectedServices: ['ws'],
        },
      ],
      generatedAt: new Date().toISOString(),
    }),
  ),
  http.get('/api/v1/status/history', ({ request }) => {
    const url = new URL(request.url);
    const slug = url.searchParams.get('service') || '';
    const days = Math.min(90, Math.max(1, Number(url.searchParams.get('days') || '30')));
    return HttpResponse.json({ service: slug, days, buckets: fakeBuckets(slug, days) });
  }),
];

File: backend/services/admin/domain/status.go
// status.go — public status page entities + repo contracts.
//
// The /status page is PUBLIC (no admin gate, no bearer auth) — it is the
// druz9 uptime/transparency surface. Two data sources:
//
//   - StatusProber.Probe — a synchronous health check against every
//     infrastructure component (Postgres, Redis, etc.). Slow components
//     (>250ms) are flagged "degraded"; failures are "down".
//   - IncidentRepo.Recent — last N rows of the `incidents` table for the
//     "Recent incidents" list.
//
// Uptime percentages are computed from the incident log: total downtime in
// the window divided by the window length, subtracted from 100%. Without a
// per-second up/down stream from Prometheus this is the best we can do
// today; a future iteration can swap in `prometheus.Range` queries here
// without touching the use case or the ports.
package domain

import (
	"context"
	"time"
)

// StatusOverall enumerates the aggregate page status.
type StatusOverall string

const (
	StatusOperational StatusOverall = "operational"
	StatusDegraded    StatusOverall = "degraded"
	StatusDown        StatusOverall = "down"
)

// StatusServiceState is the per-component health row.
type StatusServiceState struct {
	Name      string
	Slug      string
	Status    StatusOverall
	Uptime30D float64 // 0.0 – 100.0
	LatencyMS int64
}

// StatusIncident mirrors an incidents row.
type StatusIncident struct {
	ID               string
	Title            string
	Description      string
	Severity         string // minor / major / critical
	StartedAt        time.Time
	EndedAt          *time.Time
	AffectedServices []string
}

// StatusPage is the assembled response.
type StatusPage struct {
	OverallStatus StatusOverall
	Uptime90D     float64
	Services      []StatusServiceState
	Incidents     []StatusIncident
	GeneratedAt   time.Time
}

// StatusProber probes every infra component and returns their per-service
// state. Implementations MUST honour the ctx deadline — the public endpoint
// caps total work at ~2s.
type StatusProber interface {
	Probe(ctx context.Context) ([]StatusServiceState, error)
}

// IncidentRepo is the read-only adapter over the incidents table.
type IncidentRepo interface {
	Recent(ctx context.Context, limit int) ([]StatusIncident, error)
	// DowntimeSeconds returns the cumulative downtime (in seconds) over
	// `window`, as a sum over (ended_at - started_at) for incidents that
	// overlap the window. Open incidents (ended_at IS NULL) are clamped to
	// `now`.
	DowntimeSeconds(ctx context.Context, window time.Duration, now time.Time) (int64, error)
	// DailyBuckets returns a per-day status series (oldest → newest) of
	// length `days`, derived from the incidents log. If `slug` is empty the
	// page-wide worst status is reported; otherwise only incidents whose
	// affected_services contains `slug` are considered.
	DailyBuckets(ctx context.Context, slug string, days int, now time.Time) ([]StatusDayBucket, error)
}

// StatusDayBucket is one day of the per-service spark history.
type StatusDayBucket struct {
	Day    time.Time     // UTC start-of-day
	Status StatusOverall // ok if no incident touched the day
}
